### ISSUE TITLE
Provide a test framework to allow unit testing

### ISSUE THREAD

**Issue by thallgren** [jirenius/go-res#68]
I would like to have an easy way of writing tests for the go-res based services that I write. From the looks of it, a lot of this is in existence already but it isn't usable from other modules due to unexported stuff. What I'm after is something similar to net/httptest, i.e. a mock service that would allow me to send request and test responses without having to start additional services.

**Comment by jirenius** [comment]
@thallgren Sorry for the slow response. My notifications must have failed.

I fully understand the purpose and need.

The testing framework found in *go-res/test* could partially be reused. It is primarily the *MockConn* in [natstest.go](https://github.com/jirenius/go-res/blob/master/test/natstest.go) that would be universally usable.

It provides a mock connection to NATS, and might even use an actual gnatsd instance internally if needed. This is required when testing query events due to unexported fields in the *nats* package.

The test package would, like httptest, be a separate subpackage: *github.com/jirenius/go-res/restest*

Some modifications are probably desired as testing the go-res package, and testing services/handler implementations for go-res will look slightly different.

### Example Usage
```go
func TestFooHandler(t *testing.T) {
  // Service/Handler we wish to test.
   s := res.Service("foo")
   s.Handle("bar.$id",
      res.Access(res.AccessGranted),
      res.GetModel(func(r res.ModelRequest) {
         r.Model(map[string]string{
            "message": r.PathParam("id"),
         })
      }),
   )
   // Create mock connection that starts the service
   // and makes is ready to receive requests.
   c := restest.NewConn(s)
   defer c.Close()

   // Send Get request to service
   req := c.Get("foo.bar.42")
   // Get the response (t.Fatal is called in case of timeout).
   // And assert the response matches the expected data
   // once marshalled to JSON.
   req.Response(t).
        AssertModel(t, map[string]string{"message": "42"})
}
```

And similar methods would be available to test other types of requests (Call, Auth, and Access). As well as validating events and responding to query events.

Any comments on the example API?

**Comment by thallgren** [comment]
The example looks great. That's more or less exactly what I need.

I actually succeeded (well at least partially) with the current framework. The primary nuisance is that I'm not able to use the channel attribute in the `test.Session` because it's private. What I did was this:
```go
func createSession(<my parameters here>, t *testing.T) (*test.Session, chan struct{}) {
  t.Helper()

  var s *test.Session
  c := test.NewTestConn(false)
  r := res.NewService("inventory")
  r.SetLogger(logger.NewMemLogger())

  s = &test.Session{
    MockConn: c,
    Service:  r,
  }
  cl := make(chan struct{}) // can't use the one in test.Session because it's not exported

  // My code creating a service and adding handlers here

  go func() {
    defer s.StopServer()
    defer close(cl)
    if err := r.Serve(c); err != nil {
      panic("test: failed to start service: " + err.Error())
    }
  }()

  s.GetMsg(t).AssertSubject(t, "system.reset")

  return s, cl
}

const timeoutDuration = 5 * time.Second

func shutdownSession(s *test.Session, cl chan struct{}) {
  err := s.Shutdown()

  // Check error, as an error means that server hasn't had
  // time to start. We can then ignore waiting for the closing
  if err == nil {
    select {
    case <-cl:
    case <-time.After(timeoutDuration):
      panic("test: failed to shutdown service: timeout")
    }
  }
}
```
A typical test looks something like this:
```go
func TestSomething(t *testing.T) {
  s, cl := createSession(<some parameters>, t)
  // actual test code using the Session goes here
  shutdownSession(s, cl)
}

**Comment by jirenius** [comment]
Resolved in #72
